contrib/language/filters/http/source/language_filter.cc

```diff
@@ -0,0 +1,162 @@
+#include "contrib/language/filters/http/source/language_filter.h"
+
+#include <string>
+
+#include "envoy/common/exception.h"
+
+#include "source/common/common/empty_string.h"
+#include "source/common/common/logger.h"
+#include "source/common/common/utility.h"
+#include "source/common/http/utility.h"
+
+#include "absl/strings/ascii.h"
+#include "contrib/envoy/extensions/filters/http/language/v3alpha/language.pb.h"
+#include "unicode/locid.h"
+#include "unicode/utypes.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Language {
+
+// Validates a string for Accept-Language header value compliance.
+// See https://httpwg.org/specs/rfc7231.html#header.accept-language
+bool validateHeaderValue(absl::string_view value_str) {
+  for (auto ch : value_str) {
+    if (!(absl::ascii_isalnum(ch) || ch == '_' || ch == '-' || ch == '=' || ch == '.' ||
+          ch == ',' || ch == ';' || ch == '*' || ch == ' ')) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+LanguageFilterConfigImpl::LanguageFilterConfigImpl(
+    const envoy::extensions::filters::http::language::v3alpha::Language& proto_config,
+    const std::string& stats_prefix, Stats::Scope& scope)
+    : stats_(LanguageFilter::generateStats(stats_prefix, scope)) {
+  icu::Locale locale = icu::Locale(proto_config.default_language().data());
+
+  if (locale.isBogus()) {
+    throw EnvoyException(
+        fmt::format("Invalid locale default_language: {}", proto_config.default_language().data()));
+  }
+
+  default_language_ = proto_config.default_language().data();
+
+  for (const auto& supported_language : proto_config.supported_languages()) {
+    icu::Locale locale = icu::Locale(supported_language.data());
+
+    if (locale.isBogus()) {
+      throw EnvoyException(
+          fmt::format("Invalid locale supported_languages: {}", supported_language.data()));
+    }
+
+    supported_languages_.push_back(icu::Locale(supported_language.data()));
+  }
+}
+
+LanguageStats& LanguageFilterConfigImpl::stats() { return stats_; }
+
+const icu::Locale& LanguageFilterConfigImpl::defaultLanguage() const { return default_language_; }
+
+const std::vector<icu::Locale>& LanguageFilterConfigImpl::supportedLanguages() const {
+  return supported_languages_;
+}
+
+LanguageFilter::LanguageFilter(const LanguageFilterConfigSharedPtr config)
+    : config_(std::move(config)) {
+  UErrorCode errorCode = U_ZERO_ERROR;
+
+  locale_matcher_ = std::make_shared<icu::LocaleMatcher>(
+      icu::LocaleMatcher::Builder()
+          .setSupportedLocales(config_->supportedLanguages().begin(),
+                               config_->supportedLanguages().end())
+          .setDefaultLocale(&config_->defaultLanguage())
+          .build(errorCode));
+
+  if (U_FAILURE(errorCode)) {
+    ENVOY_LOG(error, "ICU error code icu::LocaleMatcher::Builder: {}", errorCode);
+
+    throw EnvoyException("Failed to initialize icu::LocaleMatcher.");
```

Comment:
Note, that this will cause abnormal process termination, since this exception will not be handled on the worker thread. You should move matcher creation to the filter config object. Note that matcher object should be readonly and possible to use concurrently.

Reply:
I'm looking into this.